ibm/mas_devops/roles/cp4d_install/tasks/main.yml

```diff
@@ -1,69 +1,64 @@
-- name: "Create CPD Meta Namespace"
+# - fail if neither cpd_version nor mas_channel provided, as we need something to determine which cpd version to be installed.
+- name: "Fail if defined cpd_version is not supported"
   vars:
-    namespace: "{{cpd_meta_namespace}}"
-  community.kubernetes.k8s:
-    apply: yes
-    definition: "{{ lookup('template', 'templates/namespace.yaml') }}"
-  register: _cpd_meta_namespace_result
-
+    cpd_supported_versions: "{{ cpd_default_supported_versions }}"
+  when:
+    - cpd_version is defined and cpd_version != ''
+    - cpd_version not in cpd_supported_versions
+  fail:
+    msg: "cpd_version {{ cpd_version }} is not supported. cpd_versions supported are {{ cpd_supported_versions }}"
 
-# 2. Setup entitlement for CP4D
+# 2. Check if CPD version is compatible with MAS version to be installed (if any)
 # -----------------------------------------------------------------------------
-- name: "Create ibm-entitlement-key secret"
-  vars:
-    entitledAuthStr: "{{cpd_registry_user}}:{{cpd_registry_password}}"
-    entitledAuth: "{{ entitledAuthStr | b64encode }}"
-    content:
-      - "{\"auths\":{\"{{cpd_registry}}\":{\"username\":\"{{cpd_registry_user}}\",\"password\":\"{{cpd_registry_password}}\",\"auth\":\"{{entitledAuth}}\"}"
-      - "}"
-      - "}"
-  community.kubernetes.k8s:
-    definition:
-      apiVersion: v1
-      kind: Secret
-      type: kubernetes.io/dockerconfigjson
-      metadata:
-        name: ibm-entitlement-key
-        namespace: "{{ cpd_meta_namespace }}"
-      stringData:
-        .dockerconfigjson: "{{ content | join('') | string }}"
+# - fail when MAS 8.7 installed not installed with cpd_version = 4.0
+# - fail when MAS 8.6 and below not installed with cpd_version = 3.5
+- name: "Fail if cpd_version not compatible with mas_channel"
+  when:
+    - mas_channel is defined
+    - mas_channel !=''
+    - (cpd_version == 'cpd35' and (mas_channel in mas87_channels)) or (cpd_version == 'cpd40' and (mas_channel not in mas87_channels))
 
+  fail:
+    msg: "cpd_version {{ cpd_version }} is not compatible with mas_channel {{ mas_channel }}"
+
+# 3. Check which CPD version to install
+# -----------------------------------------------------------------------------
+# - if cpd_version not provided, then use mas_channel to determine which cpd version is compatible
+- name: Check CP4D version to be installed
+  when:
+    - cpd_version is not defined or cpd_version == ''
+  set_fact:
+    cpd_version: "{{ 'cpd40' if (mas_channel in mas87_channels) else 'cpd35' }}"
 
-# 3. Install CP4D operator
+# 4. Check if CPD version is compatible with cluster type
 # -----------------------------------------------------------------------------
-- name: "Install CPD Subscription"
-  community.kubernetes.k8s:
-    apply: yes
-    definition: "{{ lookup('template', 'templates/subscription.yaml') }}"
-  register: _cpd_subscription_result
+# - fail if cluster_type = fyre/quickburn and cpd_version = cpd40
```

Comment:
Can we take this out please .. let's not codify current limitations based on cluster type .. there's no reason the infrastructure provider should impact this ... and presumably this answers my earlier question about why the cluster_type and cluster_name had leaked into something that shouldn't be dealing with such things.

Reply:
Yes, that's right - as I mentioned above, we're limited for now to install cpd v4 with this code on ROKS.
I can remove it, but if someone tries to run this against a non-roks, then 'cpd_install' role will fail here:
https://github.com/ibm-mas/ansible-devops/blob/manage-components-cpd40/ibm/mas_devops/roles/cp4d_install/tasks/cpd40/cpd-install.yml#L42

Should I simply remove this validation/check? Any suggestions?

Reply:
Exactly, i was surprised when I saw that in the steps... ok, will keep it then.
Here are the cpd40 docs explaining this process if you're interested: 
https://www.ibm.com/docs/en/cloud-paks/cp-data/4.0?topic=tasks-configuring-your-cluster-pull-images